File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Solution {
    private static final char MAX = '1';
    private static final char INSERT = 'I';
    private static final String DELIM = " ";

    private final Node MAX_NIL = new Node(Integer.MIN_VALUE >> 1);
    private final Node MIN_NIL = new Node(Integer.MAX_VALUE >>> 1);

    private final class Node implements Comparable<Node> {
        boolean del;
        int val;

        Node(int val) {
            this.val = val;
        }

        @Override
        public int compareTo(Node o) {
            return o.val - val;
        }
    }

    public int[] solution(String[] operations) {
        int i;
        int len;
        int[] ans;
        Node node;
        StringTokenizer st;
        PriorityQueue<Node> maxPq;
        PriorityQueue<Node> minPq;

        len = operations.length;
        maxPq = new PriorityQueue<>(len + 1);
        minPq = new PriorityQueue<>(len + 1, (o1, o2) -> o1.val - o2.val);
        maxPq.add(MAX_NIL);
        minPq.add(MIN_NIL);
        for (i = 0; i < len; i++) {
            st = new StringTokenizer(operations[i], DELIM, false);
            if (st.nextToken().charAt(0) == INSERT) {
                node = new Node(Integer.parseInt(st.nextToken()));
                minPq.offer(node);
                maxPq.offer(node);
            } else if (st.nextToken().charAt(0) == MAX) {
                do {
                    node = maxPq.poll();
                } while (node.del);
                if (node == MAX_NIL) {
                    maxPq.offer(node);
                } else {
                    node.del = true;
                }
            } else {
                do {
                    node = minPq.poll();
                } while (node.del);
                if (node == MIN_NIL) {
                    minPq.offer(node);
                } else {
                    node.del = true;
                }
            }
        }
        ans = new int[2];
        do {
            node = maxPq.poll();
        } while (node.del);
        if (node == MAX_NIL) {
            return ans;
        }
        ans[0] = node.val;
        do {
            node = minPq.poll();
        } while (node.del);
        ans[1] = node.val;
        return ans;
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cpp
#include <string>
#include <vector>
#include <queue>
#include <functional>
#include <sstream>
#include <climits>

using namespace std;

constexpr char INS = 'I';
constexpr int MAX = 1;

struct Node {
    bool del;
    int num;

    Node(int num) : del(false), num(num) {}
};

struct max_cmp {
    bool operator()(Node* a, Node* b) const {
        return a->num < b->num;
    }
};

struct min_cmp {
    bool operator()(Node* a, Node* b) const {
        return a->num > b->num;
    }
};

vector<int> solution(vector<string> operations) {
    char op;
    int i;
    int len;
    int num;
    Node* node;
    Node* max_nil;
    Node* min_nil;
    vector<int> ans;
    priority_queue<Node*, vector<Node*>, max_cmp> max_pq;
    priority_queue<Node*, vector<Node*>, min_cmp> min_pq;
    stringstream ss;

    max_pq.emplace(max_nil = new Node(INT_MIN));
    min_pq.emplace(min_nil = new Node(INT_MAX));
    len = operations.size();
    for (i = 0; i < len; i++) {
        ss.clear();
        ss.str(operations[i]);
        ss >> op;
        ss >> num;
        if (op == INS) {
            node = new Node(num);
            min_pq.emplace(node);
            max_pq.emplace(node);
        } else {
            if (num == MAX) {
                while ((node = max_pq.top())->del) {
                    max_pq.pop();
                }
                if (node != max_nil) {
                    node->del = true;
                    max_pq.pop();
                }
            } else {
                while ((node = min_pq.top())->del) {
                    min_pq.pop();
                }
                if (node != min_nil) {
                    node->del = true;
                    min_pq.pop();
                }
            }
        }
    }
    while ((node = max_pq.top())->del) {
        max_pq.pop();
    }
    if (node == max_nil) {
        ans.emplace_back(0);
        ans.emplace_back(0);
    } else {
        ans.emplace_back(node->num);
        while ((node = min_pq.top())->del) {
            min_pq.pop();
        }
        ans.emplace_back(node->num);
    }
    return ans;
}